src/v0/destinations/intercom/util.js

```diff
@@ -1,12 +1,215 @@
+  'Content-Type': JSON_MIME_TYPE,
+  Authorization: `Bearer ${destination.Config.apiKey}`,
+  Accept: JSON_MIME_TYPE,
+  'Intercom-Version': '2.10',
+});
+
+/**
+ * Returns base endpoint
+ * @param {*} destination
+ * @returns
+ */
+const getBaseEndpoint = (destination) => {
+  const { apiServer } = destination.Config;
+  if (apiServer === 'eu') {
+    return BASE_EU_ENDPOINT;
+  }
+  if (apiServer === 'au') {
+    return BASE_AU_ENDPOINT;
+  }
+  return BASE_ENDPOINT;
+};
+
+/**
+ * Returns custom attributes for identify and group calls
+ * @param {*} payload
+ * @param {*} ReservedAttributes
+ * @returns
+ */
+const filterCustomAttributes = (payload, ReservedAttributes) => {
+  const { custom_attributes } = payload;
+  let customAttributes = cloneDeep(custom_attributes) || {};
+  if (customAttributes) {
+    ReservedAttributes.forEach((trait) => {
+      delete customAttributes[trait];
+    });
+  }
+  if (isDefinedAndNotNull(customAttributes) && Object.keys(customAttributes).length > 0) {
+    customAttributes = flattenJson(customAttributes);
+  }
+  return customAttributes;
+};
+
+/**
+ * Returns transformed payload
+ * @param {*} message
+ * @param {*} category
+ * @returns
+ */
+const getPayload = (message, category) => {
+  let payload;
+  if (get(message, MappedToDestinationKey)) {
+    addExternalIdToTraits(message);
+    payload = getFieldValueFromMessage(message, 'traits');
+  } else {
+    payload = constructPayload(message, MappingConfig[category.name]);
+  }
+  return payload;
+};
+
+/**
+ * Returns lookupField
+ * @param {*} message
+ * @returns
+ */
+const getLookUpField = (message) => {
+  const integrationsObj = getIntegrationsObj(message, 'INTERCOM');
+  let lookupField = 'email';
+  if (integrationsObj && integrationsObj.lookup && isDefinedAndNotNull(integrationsObj.lookup)) {
+    lookupField = integrationsObj.lookup;
+  }
+  return lookupField;
+};
+
+/**
+ * Returns contact id based on lookup
+ * @param {*} message
+ * @param {*} destination
+ * @param {*} lookupField
+ * @returns
+ */
+const fetchContactId = async (message, destination, lookupField) => {
+  const lookupFieldValue = getFieldValueFromMessage(message, lookupField);
+  const data = JSON.stringify({
+    query: {
+      operator: 'AND',
+      value: [
+        {
+          field: lookupField,
+          operator: '=',
+          value: lookupFieldValue,
+        },
+      ],
+    },
+  });
+
+  const headers = getHeaders(destination);
+  const baseEndPoint = getBaseEndpoint(destination);
+  const endpoint = `${baseEndPoint}/${SEARCH_CONTACT_ENDPOINT}`;
+  const response = await httpPOST(endpoint, data, {
+    headers,
+    destType: 'intercom',
+    feature: 'transformation',
+  });
+
+  const processedUserResponse = processAxiosResponse(response);
+  if (isHttpStatusSuccess(processedUserResponse.status)) {
+    return processedUserResponse.response?.data.length > 0
+      ? processedUserResponse.response?.data[0]?.id
+      : null;
+  }
+
+  return null;
+};
```

Comment:
We need to do 1 search api call for the group of events we are getting in router transform to make it efficient and pass the state to transformation